OCPBUGS-33787: Tolerate the absence of ingress capability on HyperShift clusters

alebedev87 opened the pull request:
Part of the implementation of [Make Ingress optional for HyperShift EP](https://github.com/openshift/enhancements/pull/1415). Note that this change targets only HyperShift managed deployments and should not impact standalone OpenShift installations.

This PR:
- Implements [alternative ingress fields](https://github.com/openshift/api/pull/1887) from the console operator config API
- Skips component route customizations if ingress capability is disabled
- Uses NodePort type for console and downloads services if ingress capability is disabled
- Adds document to describe how to implement alternative ingress on ROSA

alebedev87 commented:
/retest

jhadvig commented on pkg/console/controllers/clidownloads/controller.go:
```diff
@@ -114,8 +121,27 @@ func (c *CLIDownloadsSyncController) Sync(ctx context.Context, controllerContext
 
 	statusHandler := status.NewStatusHandler(c.operatorClient)
+
+	infrastructureConfig, err := c.infrastructureClient.Get(ctx, api.ConfigResourceName, metav1.GetOptions{})
```
If the ingress can be enabled/disabled during the clusters lifecycle then we need to keep the fetching logic per controller, but we need to use listers rather then client.
If the ingress can be re-enabled then please move the fetching logic `starter.go`  and pass the config to each of the controllers.

alebedev87 replied:
On the standalone OpenShift the Ingress capability [will always be enabled](https://github.com/openshift/cluster-version-operator/pull/946). On HyperShift: once enabled it [cannot be disabled](https://github.com/openshift/enhancements/blob/master/enhancements/installer/component-selection.md#capabilities-cannot-be-uninstalled).

jhadvig commented on pkg/console/controllers/route/controller.go:
```diff
@@ -116,8 +124,26 @@ func (c *RouteSyncController) Sync(ctx context.Context, controllerContext factor
+	}
+
+	// Disable the route check for external control plane topology (hypershift) if the ingress capability is disabled.
+	if util.IsExternalControlPlaneWithIngressDisabled(infrastructureConfig, clusterVersionConfig) {
```
by doing this you are basically disabling the operator from doing its work by syncing the console routes.

alebedev87 replied:
@jhadvig: From what I see both of the functions used by the route controller (`SyncCustomRoute`,`SyncDefaultRoute`) are related to the component route implementation (custom hostname or tls). The component routes are not supported on HyperShift, so I think we are safe to skip them.

@csrwng: can you please confirm that the component routes are not supported on HyperShift?

csrwng replied:
That is correct. If you've disabled ingress you are also indirectly disabling component routes, no?

alebedev87 replied:
Right, I put my followup comment into a wrong conversation. Here it is:

Also, in the absence of the ingress operator the required RBAC won't be created for the components (including the console operator): [Component routes chapter of the EP](https://github.com/openshift/enhancements/blob/master/enhancements/ingress/optional-ingress-hypershift.md#component-routes).

jhadvig commented on pkg/console/controllers/oauthclients/oauthclients.go:
```diff
@@ -108,6 +116,24 @@ func (c *oauthClientsController) sync(ctx context.Context, controllerContext fac
+
+	// Disable the oauth client update for external control plane topology (hypershift) if the ingress capability is disabled.
+	// HyperShift handles the updating of the oauth client for the console.
+	if util.IsExternalControlPlaneWithIngressDisabled(infrastructureConfig, clusterVersionConfig) {
```
by doing this you are basically disabling the operator from doing its work by syncing the console oauthclients.

alebedev87 replied:
By "syncing the console oauthclients" you mean adding `redirectURI`? From what I understand this part will be handled by HyperShift as it takes care of some of the `OAuthClient`s already. I suppose it's because [the authentication server is not exposed via a route](https://github.com/openshift/enhancements/blob/master/enhancements/ingress/optional-ingress-hypershift.md#authentication-operator).

@csrwng:  What do you think? Should be disable the updates of console's `OAuthClient` on HyperShift or we can keep it for the time being?

alebedev87 replied:
Also, in the absence of the ingress operator the required RBAC won't be created for the components (including the console operator): [Component routes chapter of the EP](https://github.com/openshift/enhancements/blob/master/enhancements/ingress/optional-ingress-hypershift.md#component-routes).

csrwng replied:
Makes sense to disable updates of the OAuthClient

alebedev87 replied:
[The new API](https://github.com/openshift/api/pull/1887) has been implemented to support an alternative ingress.

jhadvig commented on pkg/console/controllers/clidownloads/controller.go:
```diff
@@ -114,8 +121,27 @@ func (c *CLIDownloadsSyncController) Sync(ctx context.Context, controllerContext
+	}
+
+	// Disable the client download check for external control plane topology (hypershift) if the ingress capability is disabled.
+	if controllersutil.IsExternalControlPlaneWithIngressDisabled(infrastructureConfig, clusterVersionConfig) {
```
by doing this you are basically disabling the operator from doing its work by syncing the console CLI downloads CRDs.

alebedev87 replied:
Right, `oc-cli-downloads` contains the links to download the `oc` client from the download route:

```bash
$ oc get consoleclidownloads oc-cli-downloads -o yaml | yq .spec.links[].href
https://downloads-openshift-console.apps.alebedev-0424.devcluster.openshift.com/amd64/linux/oc.tar
https://downloads-openshift-console.apps.alebedev-0424.devcluster.openshift.com/amd64/mac/oc.zip
https://downloads-openshift-console.apps.alebedev-0424.devcluster.openshift.com/amd64/windows/oc.zip
https://downloads-openshift-console.apps.alebedev-0424.devcluster.openshift.com/arm64/linux/oc.tar
https://downloads-openshift-console.apps.alebedev-0424.devcluster.openshift.com/arm64/mac/oc.zip
https://downloads-openshift-console.apps.alebedev-0424.devcluster.openshift.com/ppc64le/linux/oc.tar
https://downloads-openshift-console.apps.alebedev-0424.devcluster.openshift.com/s390x/linux/oc.tar
https://downloads-openshift-console.apps.alebedev-0424.devcluster.openshift.com/oc-license
```

So, the risk is to have the broken links in `/command-line-tools` path of the console UI.

alebedev87 replied:
@csrwng : can we accept the risk I described above?

csrwng replied:
Yes

jhadvig replied:
Then we should be also prematurely exiting the DownloadsDeployment controller and [remove the deployment itself](https://github.com/openshift/console-operator/blob/master/pkg/console/controllers/downloadsdeployment/controller.go#L129).

alebedev87 replied:
[The new API](https://github.com/openshift/api/pull/1887) has been implemented to setup alternative ingress for the downloads.

jhadvig reviewed (changes requested):
Thanks @alebedev87 
This change does not make the operator to go downgraded when the ingress is disabled but whats basically because you are prematurely exiting all of the updated controllers. I doubt that the console will be working in this state.
Have you tested it ?

alebedev87 commented:
>  I doubt that the console will be working in this state. Have you tested it ?

Yes. You are right, the console is not not working out of the box because the routes are left unserved. I was playing with [AWS Load Balancer Operator](https://github.com/openshift/enhancements/blob/master/enhancements/ingress/optional-ingress-hypershift.md#setup-aws-alb-as-the-new-ingress) (ALBO) to find the right `Ingress` configuration to make the console work without routes. I manged to find it but it sill has some gaps:
- the console service should be of type `NodePort` to work with the AWS ALB. This can be treated as a gap in the console operator which we can address later (on demand).
- the redirect to the oauth server (at the login time) still uses the oauth route despite the fact that I changed all `OAuth` clients to use the AWS ALBs provisioned by the aws-lb-controller. This part may be configurable though, I just me miss the console knowledge.

alebedev87 commented:
/retest

alebedev87 commented:
Rebased from `master` to fix the merge conflict.

alebedev87 commented:
/assign @jhadvig

alebedev87 commented:
/retest

jhadvig commented:
>Yes. You are right, the console is not not working out of the box because the routes are left unserved. I was playing with [AWS Load Balancer Operator](https://github.com/openshift/enhancements/blob/master/enhancements/ingress/optional-ingress-hypershift.md#setup-aws-alb-as-the-new-ingress) (ALBO) to find the right Ingress configuration to make the console work without routes. I managed to find it but it sill has some gaps:
> - the console service should be of type NodePort to work with the AWS ALB. This can be treated as a gap in the console operator which we can address later (on demand).

So this case will need to be handle in the service controller, which will then update the service spec accordingly. If I understand it, console itself wont be available though the service if the NodePort is not set? If thats the case I think it needs to be address in this change.

> - the redirect to the oauth server (at the login time) still uses the oauth route despite the fact that I changed all OAuth clients to use the AWS ALBs provisioned by the aws-lb-controller. This part may be configurable though, I may just miss some console knowledge.

Are these steps documented somewhere?

I feel like there is too many gaps - in order to merge this change. What would be the benefit for Hypershift user if he disables the Ingress capability, but the console wont be working? I feel that at minimum we should have a documented way how to make the console work.
Also this will need to be tested with QE before merging and they will need to know the steps how to:
1. setup the env (hypershift with disabled ingress capability)
after which they will test the non-functional console. In which case they will assume that its a bug. Therefor they will need a reproduction steps how to make console work.

spadgett commented:
> > I doubt that the console will be working in this state. Have you tested it ?
> 
> Yes. You are right, the console is not not working out of the box because the routes are left unserved. I was playing with [AWS Load Balancer Operator](https://github.com/openshift/enhancements/blob/master/enhancements/ingress/optional-ingress-hypershift.md#setup-aws-alb-as-the-new-ingress) (ALBO) to find the right `Ingress` configuration to make the console work without routes. I managed to find it but it sill has some gaps:
> 
> * the console service should be of type `NodePort` to work with the AWS ALB. This can be treated as a gap in the console operator which we can address later (on demand).
> * the redirect to the oauth server (at the login time) still uses the oauth route despite the fact that I changed all `OAuth` clients to use the AWS ALBs provisioned by the aws-lb-controller. This part may be configurable though, I may just miss some console knowledge.

Should we simply disable console for now if we know it won't work, even with manual load balancer configuration?

The console reads the OAuth URLs from the well-known endpoint. If the OAuth metadata can be updated in that document, console should discover the right URL.

https://docs.openshift.com/container-platform/4.15/authentication/configuring-internal-oauth.html#oauth-server-metadata_configuring-internal-oauth

alebedev87 commented:
> Should we simply disable console for now if we know it won't work, even with manual load balancer configuration?

The console is still an important part of HyperShift based offerings like ROSA. The HyperShift engineers prefer the console to be running even without the default ingress (router). The way to expose it is a subject to a discussion, one of the ways to expose the console is described in [the doc I added today](https://github.com/openshift/console-operator/blob/ca595bff46855c31fc4c6204f13c726d9eaa5fbd/docs/alb-ingress-rosa-hcp.md).
However the console operator depends deeply on the router being present on the cluster. And we need to think pragmatically about what we can do in 4.16 release and what needs a follow up.

> The console reads the OAuth URLs from the well-known endpoint. If the OAuth metadata can be updated in that document, console should discover the right URL.

ROSA HCP clusters already have [the auth URL which is not an OpenShift route](https://github.com/openshift/console-operator/blob/ca595bff46855c31fc4c6204f13c726d9eaa5fbd/docs/alb-ingress-rosa-hcp.md#notes). This makes the authentication from console work smoothly even without the router.

spadgett commented:
> The console is still an important part of HyperShift based offerings like ROSA. The HyperShift engineers prefer the console to be running even without the default ingress (router). The way to expose it is a subject to a discussion, one of the ways to expose the console is described in [the doc I added today](https://github.com/openshift/console-operator/blob/ca595bff46855c31fc4c6204f13c726d9eaa5fbd/docs/alb-ingress-rosa-hcp.md). 

But I don't think this will work since the console will stomp those changes, unless you put it in unmanaged state.

> > The console reads the OAuth URLs from the well-known endpoint. If the OAuth metadata can be updated in that document, console should discover the right URL.
> 
> ROSA HCP clusters already have [the auth URL which is not an OpenShift route](https://github.com/openshift/console-operator/blob/ca595bff46855c31fc4c6204f13c726d9eaa5fbd/docs/alb-ingress-rosa-hcp.md#notes). This makes the authentication from console work smoothly even without the router.

That console doesn't really care whether it's a route or not. It just needs to be exposed in the well-known endpoint so it discovers the right URL. (We might need to double check how console gets the CA cert, however.)

alebedev87 commented:
Commit added to let the oauthclient controller retrieve the ingress hostname from console config if the ingress capability is disabled.

alebedev87 commented:
/jira refresh

alebedev87 commented:
> > The console is still an important part of HyperShift based offerings like ROSA. The HyperShift engineers prefer the console to be running even without the default ingress (router). The way to expose it is a subject to a discussion, one of the ways to expose the console is described in [the doc I added today](https://github.com/openshift/console-operator/blob/ca595bff46855c31fc4c6204f13c726d9eaa5fbd/docs/alb-ingress-rosa-hcp.md).
> 
> But I don't think this will work since the console will stomp those changes, unless you put it in unmanaged state.

@spadgett: In this PR I modified the console operator to _not_ stomp on the console config and oauthclient if the ingress capability is disabled on external control plane topology (HyperShift). This allows the end user can provide it's own console base address (ALB is an example). I designed the way to provide the console base address too - via the same console config. WDYT?

spadgett commented:
> In this PR I modified the console operator to not stomp on the console config and oauthclient if the ingress capability is disabled on external control plane topology (HyperShift).

Are you referring to the internal console config ConfigMap managed by the operator or the public console operator config resource? Should we introduce the property to the console operator config itself? Then we could have proper API for specifying the public URLs when ingress is disabled.

Similarly, for auth, can we update the well-known endpoints with the right URLs if possible? Then console will discover the right values, along with anything else that reads the well-known document. With your current approach, I assume other components on the cluster won't be able to use auth? (For instance, anything that uses the OAuth proxy?)

How does this interact with external OIDC?

In general, it would be good to have public APIs for this configuration rather than asking customers to change internal config resources. I worry that if we change something in the internal console config in a future release, it won't get updated since console isn't stomping the changes, causing things to break.

Adding @stlaz for input.

spadgett commented:
Note that it is possible to manually edit these resources today without code changes if you put the console in `Unmanaged` state (but that will block cluster upgrades).

alebedev87 commented:
> Are you referring to the internal console config ConfigMap managed by the operator or the public console operator config resource?

Internal config, `console-config` configmap.

> Should we introduce the property to the console operator config itself? Then we could have proper API for specifying the public URLs when ingress is disabled.

I chose the internal config because it already had the console base address field. Yes, the proper API may be a way to go. My only concern is time, this feature was promised in 4.14 and all the pieces are merged except for the console operator. 

> Similarly, for auth, can we update the well-known endpoints with the right URLs if possible?

It's not needed on HyperShift managed clusters. The auth server is managed centrally by HyperShift and it's not exposed via a route on the user cluster.

> Then console will discover the right values, along with anything else that reads the well-known document.

On HyperShift managed cluster the console already discovers the right value for the auth server.

```
$ oc -n openshift-console rsh deploy/console curl -k https://openshift.default.svc/.well-known/oauth-authorization-server
{
"issuer": "https://oauth.mytestcluster.5199.s3.devshift.org:443",
"authorization_endpoint": "https://oauth.mytestcluster.5199.s3.devshift.org:443/oauth/authorize",
"token_endpoint": "https://oauth.mytestcluster.5199.s3.devshift.org:443/oauth/token",
```

> How does this interact with external OIDC? Will this and external OIDC ever be configured together?

I don't know. I miss the knowledge here

> I worry that if we change something in the internal console config in a future release, it won't get updated since console isn't stomping the changes, causing things to break.

The way I implemented it only the `consoleBaseAddress` field is not stomped.

alebedev87 commented:
@spadgett: talking about the public API, I can think about something like this: https://github.com/openshift/api/pull/1887.

spadgett commented:
> On HyperShift managed cluster the console already discovers the right value for the auth server.

OK, is the change that's needed then to set the right OAuth callback URL in the OAuthClient?

> The way I implemented it only the `consoleBaseAddress` field is not stomped.

This makes me feel better, but I still think a proper API would be better long term.

alebedev87 commented:
>> On HyperShift managed cluster the console already discovers the right value for the auth server.

> OK, is the change that's needed then to set the right OAuth callback URL in the OAuthClient

In this PR I set the callback URL in console OAuthClient yes. But it's set with the console base address, not the auth server. All auth URL related resources are managed elsewhere (supposedly on HyperShift management side).

alebedev87 commented:
The PR updated to use [the new field from the operator config](https://github.com/openshift/api/pull/1887).

alebedev87 commented:
/retest

melvinjoseph86 commented:
/test e2e-aws-console

alebedev87 commented:
/retest

Seems like there are more test failures after the listers came back to `master`. Retrying to see if it's not transient.

alebedev87 commented:
/jira refresh

jhadvig commented on pkg/console/subresource/configmap/configmap.go:
```diff
@@ -47,12 +47,13 @@ func DefaultConfigMap(
 	nodeOperatingSystems []string,
 	copiedCSVsDisabled bool,
 	telemeterConfig map[string]string,
+	consoleHost string,
```
I see the point of adding this additional param, but I would rather avoid it, since for the non ALB setup, `consoleHost` will be the same as the host int the `activeConsoleRoute.spec`.
Lets rather create a new struct, which would include both of these types.

alebedev87 replied:
> since for the non ALB setup, consoleHost will be the same as the host int the activeConsoleRoute.spec

Well, most of the time it's true. The `route.spec.host` is assigned automatically by the apiserver from the default ingress domain without the knowledge about the inrgesscontrollers which is only present in the route's status. That's why in `operator/sync_v400.go` I pass down the url returned by `routesub.GetActiveRouteInfo()` which returns only the admitted host.

jhadvig reviewed (commented):
Thanks a lot @alebedev87 for the PR. Looks good.
Adding couple of comments. PTAL

alebedev87 commented on docs/alb-ingress-rosa-hcp.md:
```diff
@@ -0,0 +1,155 @@
+
+3. When the ingress capability is disabled, the console operator skips the implementation of the component route customization.
+
+4. To simulate a disabled ingress set the desired replicas to zero in the default ingress controller:
```
No, the routes will still be admitted. I rephrased it to:

```
To simulate the absence of ingress connectivity when the ingress capability is disabled, set the desired replicas to zero in the default ingress controller
```

alebedev87 commented on docs/alb-ingress-rosa-hcp.md:
```diff
@@ -0,0 +1,155 @@
+
+### Create Ingress resources for the services
+
+To provision ALBs create the following resources:
```
> Wondering if this could be also done automatically by the console-operator if ALB setup is detected?

So far, we didn't see much automation around the certificate creation from the ALBO users. Probably because this requires quite some some effort to manage the cloud credentials. We can think about it as a followup enhancement. .

alebedev87 commented:
Always apply nodeport services, to avoid the chicken and egg problem with alternative ingress: `Ingress` needs a nodeport service before it can provision a LB whose host we need to configure `console.operator.spec.ingress`.

alebedev87 commented:
/retest

alebedev87 commented:
`ConsoleURL` removed from configmap config as it's not a customization but API.

alebedev87 commented:
As discussed with @jhadvig: create `NodePort` service only for clusters with disabled ingress capability.

alebedev87 commented:
/retest

alebedev87 commented:
/retest

CI outage.

alebedev87 commented:
`e2e-aws-console` test is unstable and needs to be overridden [like we did for #907](https://github.com/openshift/console-operator/pull/907#issuecomment-2144718594).

jhadvig commented:
Thanks @alebedev87 for the PR. LGTM
/lgtm
/retest

jhadvig commented:
Overriding the TC due to flaking e2e test
/override ci/prow/e2e-aws-console
